### DISCUSSION TITLE
Is it possible to use this without Winit on the web?

### DISCUSSION THREAD

**Discussion by breqdev** [parasyte/pixels#328] [Q&A] [answered]
I've been working on an emulator project targeting WASM. Currently, I'm using a simple 2D JavaScript canvas to draw the screen, but I'm running into performance issues. This library looks like it could be a much more performant approach, but I don't really want to use Winit in my project -- I'd rather handle the event loop, key presses, etc. myself. Is there a way to use pixels without using Winit, such as directly attaching the SurfaceTexture to a `<canvas>` element somehow?

**Comment by parasyte** [comment, marked as answer]
At present you need something that implements [`HasRawWindowHandle`](https://docs.rs/raw-window-handle/latest/raw_window_handle/trait.HasRawWindowHandle.html) to gain access to a WebGL (or WebGPU) context. This is by design. See https://github.com/parasyte/pixels/issues/309#issuecomment-1264441057 for rationale. This means you do not need `winit`, but you still need some abstraction layer.

If we decide to support canvases directly, it will be behind a feature flag and will pretty much be a bespoke substitute for https://github.com/rust-windowing/raw-window-handle/issues/102 (which is at an impasse and not expected to see any progress in the near future).

**Comment by breqdev** [comment]
Thank you for the fast reply! For anyone else stumbling upon this problem, here's a quick and dirty implementor for `HasRawWindowHandle` and `HasRawDisplayHandle` that worked for me:

```rust
use pixels::raw_window_handle::{
  HasRawDisplayHandle, HasRawWindowHandle, RawDisplayHandle, RawWindowHandle, WebDisplayHandle,
  WebWindowHandle,
};
use web_sys::HtmlCanvasElement;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct CanvasWindow {
  id: u32,
}

impl CanvasWindow {
  pub fn new(canvas: &HtmlCanvasElement) -> Self {
    // Get the ID from the canvas's `data-raw-handle` attribute.
    let id = canvas
      .attributes()
      .get_named_item("data-raw-handle")
      .expect("Canvas has no `data-raw-handle` attribute")
      .value()
      .parse::<u32>()
      .expect("Canvas `data-raw-handle` attribute is not a number");

    Self { id }
  }
}

unsafe impl HasRawWindowHandle for CanvasWindow {
  fn raw_window_handle(&self) -> RawWindowHandle {
    let mut handle = WebWindowHandle::empty();
    handle.id = self.id;

    RawWindowHandle::Web(handle)
  }
}

unsafe impl HasRawDisplayHandle for CanvasWindow {
  fn raw_display_handle(&self) -> RawDisplayHandle {
    let handle = WebDisplayHandle::empty();

    RawDisplayHandle::Web(handle)
  }
}
```
